Sources/NIOSOCKS/Channel Handlers/SOCKSServerHandshakeHandler.swift

```diff
@@ -0,0 +1,120 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Add this handshake handler to the front of your channel, closest to the network.
+/// The handler will receive bytes from the network and run them through a state machine
+/// and parser to enforce SOCKSv5 protocol correctness. Inbound bytes will by parsed into
+/// `ClientMessage` for downstream consumption. Send `ServerMessage` to this
+/// handler.
+public final class SOCKSServerHandshakeHandler: ChannelDuplexHandler, RemovableChannelHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ClientMessage
+    public typealias OutboundIn = ServerMessage
+    public typealias OutboundOut = ByteBuffer
+    
+    var inboundBuffer: ByteBuffer?
+    var stateMachine: ServerStateMachine
+    
+    public init() {
+        self.stateMachine = ServerStateMachine()
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        if self.stateMachine.proxyEstablished {
+            context.fireChannelRead(data)
```

Comment:
This is incompatible with the way we unbuffer data in `handlerRemoved`: if we get a `channelRead` before `handlerRemoved` fires, we're in trouble. We should probably continue to buffer until we're removed and deliver all the bytes at that point.